lib/token-store/index.js

```diff
@@ -0,0 +1,594 @@
+/**
+ * @fileoverview Object to handle access and retrieval of tokens.
+ * @author Brandon Mills
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const assert = require("assert");
+const cursors = require("./cursors");
+const ForwardTokenCursor = require("./forward-token-cursor");
+const PaddedTokenCursor = require("./padded-token-cursor");
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+const PUBLIC_METHODS = Object.freeze([
+    "getTokenByRangeStart",
+
+    "getFirstToken",
+    "getLastToken",
+    "getTokenBefore",
+    "getTokenAfter",
+    "getFirstTokenBetween",
+    "getLastTokenBetween",
+
+    "getFirstTokens",
+    "getLastTokens",
+    "getTokensBefore",
+    "getTokensAfter",
+    "getFirstTokensBetween",
+    "getLastTokensBetween",
+
+    "getTokens",
+    "getTokensBetween",
+
+    "getTokenOrCommentBefore",
+    "getTokenOrCommentAfter"
+]);
+
+/**
+ * Creates the map from locations to indices in `tokens`.
+ *
+ * The first/last location of tokens is mapped to the index of the token.
+ * The first/last location of comments is mapped to the index of the previous token of each comment.
+ *
+ * @param {Token[]} tokens - The array of tokens.
+ * @param {Comment[]} comments - The array of comments.
+ * @returns {Object} The map from locations to indices in `tokens`.
+ * @private
+ */
+function createIndexMap(tokens, comments) {
+    const map = Object.create(null);
+    let it = 0;
+    let ic = 0;
```

Comment:
Could these variables be renamed (e.g., tokenIndex, commentIndex)? (I think this came up earlier)